engine/internal/pkg/discovery/agent.go

```diff
@@ -38,182 +35,166 @@ const (
-	)
+	eventNotifier := notifier.NewNotifier[Event]()
+	defer eventNotifier.Close()
 
-	discoveryCtx, cancel := context.WithCancel(ctx)
-	defer cancel()
+	syncErr := make(chan error, 1)
+	syncDone := make(chan struct{})
+	go func() {
+		defer close(syncDone)
+		syncErr <- a.autoSync(ctx)
+	}()
 
-	// FIXME The semantics of ResetDiscovery is confusing. We will deal with this later.
-	startTime := time.Now()
-	session, err := runner.ResetDiscovery(discoveryCtx, true)
-	duration := time.Since(startTime)
-	if err != nil {
-		log.Warn("establish discovery session failed",
-			zap.Duration("duration", duration), logutil.ShortError(err))
-		return errors.Trace(err)
-	}
-	log.Info("discovery session established", zap.Duration("duration", duration))
 	defer func() {
-		_ = session.Close()
+		<-syncDone
+		log.Info("discovery agent exited")
 	}()
 
-	eventNotifier := notifier.NewNotifier[Event]()
-	defer eventNotifier.Close()
-
+	snap := make(Snapshot)
 	for {
 		select {
-		case <-discoveryCtx.Done():
-			return errors.Trace(discoveryCtx.Err())
-		case <-session.Done():
-			log.Warn("metastore session is done", zap.String("executor-id", string(a.selfInfo.ID)))
-
-			startTime := time.Now()
-			session, err = runner.ResetDiscovery(ctx, true /* resetSession*/)
-			duration := time.Since(startTime)
-
-			if err != nil {
-				log.Warn("reset discovery session failed",
-					zap.Duration("duration", duration),
-					logutil.ShortError(err))
-				return errors.Annotate(err, "discovery agent session done")
-			}
-			log.Info("reset discovery session", zap.Duration("duration", duration))
+		case <-ctx.Done():
+			return errors.Trace(ctx.Err())
+		case event := <-a.eventCh:
+			eventNotifier.Notify(event)
+			applyEvent(snap, event)
 		case req := <-a.subscribeCh:
-			req.snap = runner.GetSnapshot()
+			req.snap = snap
+			if err := eventNotifier.Flush(ctx); err != nil {
+				return errors.Trace(err)
+			}
 			req.receiver = eventNotifier.NewReceiver()
 			close(req.doneCh)
-		case resp := <-runner.GetWatcher():
-			if err := a.handleWatchResp(resp, runner, eventNotifier); err != nil {
-				return err
-			}
+		case err := <-syncErr:
+			return errors.Trace(err)
 		}
 	}
 }
 
-func (a *agentImpl) handleWatchResp(
-	resp srvdiscovery.WatchResp,
-	runner srvdiscovery.DiscoveryRunner,
-	eventNotifier *notifier.Notifier[Event],
-) error {
-	startTime := time.Now()
-	defer func() {
-		duration := time.Since(startTime)
-		if duration > 1*time.Second {
-			// Print a warning to indicate that there might be a
-			// congestion problem.
-			log.Warn("discovery agent broadcast took too long",
-				zap.Duration("duration", duration))
+func (a *agentImpl) autoSync(ctx context.Context) error {
+	snap := make(Snapshot)
+	for {
+		newSnap, err := a.getSnapshot(ctx)
+		if err != nil {
+			if errors.Cause(err) == context.Canceled || errors.Cause(err) == context.DeadlineExceeded {
+				return errors.Trace(err)
+			}
+			log.Warn("failed to get snapshot", zap.Error(err))
```

Comment:
When can this happen?

Reply:
When encountering network error or servermaster failover.